### PR TITLE
Restructure editor user interface

### LINKED ISSUE #331: Add 'pin'/'favorite' button in the note editor
Currently, it is possible to 'pin' a note in the note history view.
This feature should also be available in the note editor itself.
It seems counter-intuitive to read a note, find it 'pin worthy' and then need to switch to the history view to actually pin it.

The easiest way to add this is probably to add a pin (or favorite star) button in the top taskbar.

This is blocked by #330

### PR DESCRIPTION
**Pull request by ErikMichelson**
### Component/Part

Editor -> UI

### Description
This PR restructures the user interface of HedgeDoc's heart component: The editor.
To separate concerns of the application and the opened document itself, a new document-bar is introduced directly below the top-bar which only contains application-specific actions from now on. The information about note creation and last edits moved into a new modal dialog "document info" that can be invoked from the document bar. The toolbar of the editor is now grouped into different sections and the settings from the status-bar of the editor will move into a new editor-settings-modal invoked from the toolbar as well.

Note that this interface is still a draft and not settled into stone yet. Feel free to add comments to this PR below.

image

image

image

### Steps

- [ ] added implementation
- [ ] added / updated tests
- [ ] added / updated documentation
- [ ] extended changelog

### Related Issue(s)
closes #331 
see #93

### CODE REVIEW

**Comment by ccoenen** [comment]
One thing I'm not clear about is, why in the second menu bar some things are left-aligned with icons and some things are right-aligned. And why is import/export/share three buttons in two places?

I would at least request that the permissions thing gets moved to the right side (where all the dropdowns are) directly next to the "X online" dropdown. These two are the only things that are dropdowns _AND_ display a state (as opposed to a label like import/export).

Another thing I'd like to ask is, what is in the "menu" dropdown? The things that _used to be_ there are now directly in the menubar, are they not?

**Comment by mrdrogdrog** [comment]
> One thing I'm not clear about is, why in the second menu bar some things are left-aligned with icons and some things are right-aligned.

It looked a bit odd, when everything was on the right, so we split the actions from the menus. 

> And why is import/export/share three buttons in two places?

We have enough space. Why shouldn't we use it?

Share is not export. If that isn't clear enough, then maybe we should rename it to "share link"?

> I would at least request that the permissions thing gets moved to the right side (where all the dropdowns are) directly next to the "X online" dropdown. These two are the only things that are dropdowns _AND_ display a state (as opposed to a label like import/export).

Good point.

> Another thing I'd like to ask is, what is in the "menu" dropdown? The things that _used to be_ there are now directly in the menubar, are they not?

In the menu "menu" is now only "delete note". We still don't know where to put this action and wanted to keep the menu until the PR is ready to review. If we don't have any good reason for keeping it, then we'll delete the menu and put the action somewhere else.

My idea was to rename this menu to "delete note", but keep the menu item. Because then you'll have to click a dropdown menu, the dropdown icon AND confirm a modal. IMHO you can't have enough clicks for such a destructing action.

**Comment by DerMolly** [comment]
Also regarding the menu menu, maybe we find other items that don't fit anything else.

**Comment by mrdrogdrog** [comment]
We changed the permission dropdown to a button, that opens a modal, because the permission system is under construction and will be changed in the future.

**Comment by ccoenen** [comment]
thanks :-)

File: public/locales/en.json [hunk 1]
```diff
@@ -219,18 +219,44 @@
...
+            "pinnedToHistory": "Pinned to history"
+        },
+        "preferences": {
+            "title": "Preferences",
+            "theme": "Theme",
+            "keymap": "Keymap",
+            "tabCharacter": "Tab character",
+            "tabSize": "Tab size",
+            "spellChecker": "Spell checking"
+        },
```
Comment (mrdrogdrog) [inline public/locales/en.json:250]: shouldn't that be under "modal" ?

File: src/components/editor/editor-window/tool-bar/editor-preferences/editor-preferences.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import { EditorPreferenceTabsize } from './editor-preference-tabsize'
+import { EditorPreferenceTheme } from './editor-preference-theme'
+
+export interface EditorSettingsButtonProps {
+  editor: CodeMirror.Editor
+  preferences: EditorConfiguration
+  onPreferencesChange: (config: EditorConfiguration) => void
+}
+
+export const EditorPreferences: React.FC<EditorSettingsButtonProps> = ({ editor, onPreferencesChange, preferences }) => {
```
Comment (mrdrogdrog) [inline src/components/editor/editor-window/tool-bar/editor-preferences/editor-preferences.tsx:18]: I don't know if it's a good idea to create 4 preference specific components, that all have similar code. They all are just a dropdown for text. I think a more elegant solution would be to create a generic "select dropdown with label" component and maintain their options and the state manipulation here.

File: src/components/landing/layout/user-avatar/user-avatar.tsx [hunk 3]
```diff
@@ -1,13 +1,15 @@
...
 export interface UserAvatarProps {
     name: string;
     photo: string;
     additionalClasses?: string;
+    showName?: boolean
 }
 
-const UserAvatar: React.FC<UserAvatarProps> = ({ name, photo, additionalClasses = '' }) => {
+const UserAvatar: React.FC<UserAvatarProps> = ({ name, photo, additionalClasses = '', showName = true }) => {
   // ToDo: add Translation Key for Avatar of ${name}
```
Comment (mrdrogdrog) [inline src/components/landing/layout/user-avatar/user-avatar.tsx:13]: Can you solve this todo? :sweat_smile:

File: src/components/editor/editor-window/tool-bar/editor-preferences/editor-preferences.tsx (hunk 2 above)
Comment (ErikMichelson) [reply src/components/editor/editor-window/tool-bar/editor-preferences/editor-preferences.tsx:15]: This was a work-in-progress state and now finally just uses a single preference-selection component.

File: public/locales/en.json (hunk 1 above)
Comment (ErikMichelson) [reply public/locales/en.json]: The translation keys were restructured. "preferences" is now under "modal".

File: src/components/landing/layout/user-avatar/user-avatar.tsx (hunk 3 above)
Comment (ErikMichelson) [reply src/components/landing/layout/user-avatar/user-avatar.tsx]: Done in 3b462281

**Comment by ccoenen** [comment]
I'd like to request an updated screenshot, if that is possible.

**Comment by ErikMichelson** [comment]
@ccoenen Feel free to play around with it here: https://hedgedoc-ui.test.fsbremen.de/n/features

**Comment by mrdrogdrog** [comment]
@ccoenen 

image
image
image

**Pull request merged**